packages/components/src/slot-fill/bubbles-virtually/observable-map.ts

```diff
@@ -0,0 +1,78 @@
+/**
+ * WordPress dependencies
+ */
+import { useMemo, useSyncExternalStore } from '@wordpress/element';
+
+export type ObservableMap< K, V > = {
+	get( name: K ): V | undefined;
+	set( name: K, value: V ): void;
+	delete( name: K ): void;
+	subscribe( name: K, listener: () => void ): () => void;
+};
+
+export function observableMap< K, V >(): ObservableMap< K, V > {
+	const map = new Map< K, V >();
+	const listeners = new Map< K, Set< () => void > >();
+
+	function callListeners( name: K ) {
+		const list = listeners.get( name );
+		if ( ! list ) {
+			return;
+		}
+		for ( const listener of list ) {
+			listener();
+		}
+	}
+
+	function unsubscribe( name: K, listener: () => void ) {
+		return () => {
+			const list = listeners.get( name );
+			if ( ! list ) {
+				return;
+			}
+
+			list.delete( listener );
+			if ( list.size === 0 ) {
+				listeners.delete( name );
+			}
+		};
+	}
+
+	return {
+		get( name ) {
+			return map.get( name );
+		},
+		set( name, value ) {
+			map.set( name, value );
+			callListeners( name );
+		},
+		delete( name ) {
+			map.delete( name );
+			callListeners( name );
+		},
+		subscribe( name, listener ) {
+			let list = listeners.get( name );
+			if ( ! list ) {
+				list = new Set();
+				listeners.set( name, list );
+			}
+			list.add( listener );
+
+			return unsubscribe( name, listener );
```

Comment:
Wouldn't it be possible to avoid having to get the list again if it's inlined? Or can it disappear/change some other way?

Reply:
Good idea, will do in a followup :+1: